### PR TITLE
Make faucet retries more persistent, don't recycle coins when errors …

### PR DESCRIPTION
**Pull request by mystenmark**
…occur

### CODE REVIEW

**Review by tnowacki — COMMENTED** [review]
Looks fine to me, but I'd really defer to @amnn or @longbowlu, as I've only really become familiar with this code this week :)

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (tnowacki) [inline crates/sui-faucet/src/faucet/simple_faucet.rs:183]: `map_err` might be helpful here, though I'm not sure how it interacts with `warn!`?

Comment (amnn) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:183]: I think it would Interact fine with map_err, but I don't know that it gains you much here.

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (amnn) [inline crates/sui-faucet/src/faucet/simple_faucet.rs:269]: This function creates a fresh transaction each time, and signs it, instead we need to create and sign the transaction once, and repeatedly send it to the quorum driver. I.e. pull everything in execute_pay_sui_txn until execute_transaction outside this loop.

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (amnn) [inline crates/sui-faucet/src/faucet/simple_faucet.rs:272]: If we get a non-transient error, we can stop (and not discard the coin) @longbowlu, could you detail what errors we should be looking for here that signal we can stop and put the coin back?

It's only if the error indicates that either faucet was left waiting for a response from fullnode or fullnode was left waiting for a response from the validator that signals we should retry.  In any other case, if we got an error, we should expect to get the same error again if we run again.

**Review by amnn — COMMENTED** [review]
The structure is there in broad strokes but there are still some changes needed to how the transaction is constructed and bailing out for certain kinds of error.

**Review by longbowlu — APPROVED** [review]

**Review by longbowlu — CHANGES_REQUESTED** [review]
fat finger

**Review by longbowlu — APPROVED** [review]
Please forgive for my churns. There are a few places we can improve but this can go to stop the bleeding

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (longbowlu) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:269]: @amnn  is right that we'd usually like to keep using the exactly same transaction. In this case, the transaction is expected to the same even though we built it multiple times, because all the params are the same. The only possible variable is object version, which SHOULDN'T change in good days.

That said, it's still better to build the tx once and send the clone, assuming faucet is the only operator of this account (it should be)

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (longbowlu) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:272]: Due to the nature of what faucet does, i think there're very few cases the error is non-transient. Some common error from FN: 
1. Timeout: e.g. when validators are busy/down, and FN is waiting as well. In this case, we should retry
2. FN gave up: FN couldn't get things right within the timeout and return this error. Faucet should retry
3. Object is equivocated: discard the coin
4. invalid tx/object: shouldn't be a client error in faucet
5. insufficient gas: probably is not possible?

The errors can be found [here](https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/quorum_driver_types.rs).
I'll be working on categorizing QD errors better soon and will make clients more clever in handling failures.

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (mystenmark) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:269]: I did consider doing this but convinced myself it would be impossible for the tx to change from one construction to the next. But if everyone is in agreement then i'll do it that way.

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (mystenmark) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:272]: I agree that we can be a lot more intelligent about handling the errors here, but I don't really have time to deal with that right now - I'd like to get in the minimal change that will stop faucet from proactively locking itself out of its own gas.

Agreed with @longbowlu that QD/TO is the place to put in the work of doing it The Right Way.

File: crates/sui-faucet/src/faucet/simple_faucet.rs
Comment (tnowacki) [reply crates/sui-faucet/src/faucet/simple_faucet.rs:183]: not much, other than like one level of indentation. Amazing, I know

**Pull request merged**
